File: src/learning_service/scripts/turtle_spawn.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import rospy
from turtlesim.srv import Spawn

def turtle_spawn():
    rospy.init_node('turtle_spawn')
    rospy.wait_for_service('/spawn')
    try:
        add_turtle = rospy.ServiceProxy('/spawn', Spawn)

        response = add_turtle(2.0, 2.0, 0.0, "turtle2")
        return response.name
    except rospy.ServiceException as e:
        # print("Service call failed: %s" % e)
        print(f"Service call failed:{e}")

if __name__ == '__main__':
    # print("Spawn turtle sucessfully [name:%s]" % (turtle_spawn()))
    print(f"Spawn turtle sucessfully [name:{turtle_spawn()}]")

File: src/learning_service/src/turtle_spawn.cpp
/**
 * 该例程将请求/spawn服务，服务数据类型turtlesim::Spawn
 */

#include <ros/ros.h>
#include <turtlesim/Spawn.h>

int main(int argc, char *argv[])
{
    ros::init(argc, argv, "turtle_spawn");

    ros::NodeHandle node;

    ros::service::waitForService("/spawn");
    ros::ServiceClient add_turtle = node.serviceClient<turtlesim::Spawn>("spawn");

    turtlesim::Spawn srv;
    srv.request.x = 2.0;
    srv.request.y = 2.0;
    srv.request.name = "turtle2";

    ROS_INFO("Call service to spwan turtle[x:%0.6f, y:%0.6f, name:%s]",
            srv.request.x, srv.request.y, srv.request.name.c_str());

    add_turtle.call(srv);

    ROS_INFO("Spawn turtle successfully [name:%s]", srv.response.name.c_str());

    return 0;
}
